src/backend/dx12/src/device.rs

```diff
@@ -2808,6 +2808,16 @@ impl d::Device<B> for Device {
             d::WaitFor::All => TRUE,
         };
         let hr = unsafe {
+            // This block handles overflow when converting to u32 and always rounds up
+            // The Vulkan specification allows to wait more than specified
+            let timeout_ms = {
+                if timeout_ns > (<u32>::max_value() as u64) * 1_000_000 {
```

Comment:
this block should be rewritten, it doesn't properly handle the overflow anyway: if `timeout_ns == <u32>::max_value() as u64 * 1_000_000`, for example, the condition will be false, but it would still overflow below.
I suggest just doing `((timeout_ns + 500_000) / 1_000_000).min(<u32>::max_value() as u64) as u32`

Reply:
yeah, min is a nice idea.
but why add `500_000` instead of `999_999`?

Reply:
no, it's not good, because `timeout_ns + anything` could fail if timeout_ns is `!0`

Reply:
on a second consideration, rounding up to highest makes more sense. Let's keep 999_999 there

Reply:
It won't overflow when `timeout_ns == <u32>::max_value() as u64 * 1_000_000`. `(4294967295000000 + 999_999) / 1_000_000` equals `4294967295`, so it's not an overflow.

And I cannot use `min`, because it would overflow:

`((timeout_ns + 500_000) / 1_000_000).min(<u32>::max_value() as u64) as u32`

This would mean that the addition runs before the overflow check by the `min` function.